cstore_writer.c

```diff
@@ -239,9 +326,10 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
 			Oid columnCollation = attributeForm->attcollation;
 
 			blockData->existsArray[blockRowIndex] = true;
-			blockData->valueArray[blockRowIndex] = DatumCopy(columnValues[columnIndex],
-															 columnTypeByValue,
-															 columnTypeLength);
+
+			SerializeSingleDatum(blockData->uncompressedDataBuffer,
+								 columnValues[columnIndex], columnTypeByValue,
+								 columnTypeLength, attributeForm->attalign);
```

Comment:
Can I explain this a bit more?

What I suggested here just moves out some lines out of the ```SerializeDatum``` so it's responsibility becomes less. The enlarge operation will be executed with the same arguments and the same number of calls as before.